packages/swapper/src/swappers/cow/cowBuildTrade/cowBuildTrade.ts

```diff
@@ -0,0 +1,139 @@
+import { getNowPlusThirtyMinutesTimestamp, getUsdRate } from '../utils/helpers/helpers'
+
+export async function cowBuildTrade(
+  deps: CowSwapperDeps,
+  input: BuildTradeInput
+): Promise<Trade<'eip155:1'>> {
+  try {
+    const { sellAsset, buyAsset, sellAmount, sellAssetAccountNumber, wallet } = input
+    const { adapter, feeAsset } = deps
+
+    const { assetReference: sellAssetErc20Address, assetNamespace: sellAssetNamespace } =
+      fromAssetId(sellAsset.assetId)
+    const { assetReference: buyAssetErc20Address, assetNamespace: buyAssetNamespace } = fromAssetId(
+      buyAsset.assetId
+    )
+
+    if (buyAssetNamespace !== 'erc20' || sellAssetNamespace !== 'erc20') {
+      throw new SwapError('[cowBuildTrade] - Both assets need to be ERC-20 to use CowSwap', {
+        code: SwapErrorTypes.UNSUPPORTED_PAIR,
+        details: { buyAssetNamespace, sellAssetNamespace }
+      })
+    }
+
+    const receiveAddress = await adapter.getAddress({ wallet })
+    const normalizedSellAmount = normalizeAmount(sellAmount)
+
+    /**
+     * /v1/quote
+     * params: {
+     * sellToken: contract address of token to sell
+     * buyToken: contractAddress of token to buy
+     * receiver: receiver address can be defaulted to "0x0000000000000000000000000000000000000000"
+     * validTo: time duration during which quote is valid (eg : 1654851610 as timestamp)
+     * appData: appData for the CowSwap quote that can be used later, can be defaulted to "0x0000000000000000000000000000000000000000000000000000000000000000"
+     * partiallyFillable: false
+     * from: sender address can be defaulted to "0x0000000000000000000000000000000000000000"
+     * kind: "sell" or "buy"
+     * sellAmountBeforeFee / buyAmountAfterFee: amount in base unit
+     * }
+     */
+    const quoteResponse: AxiosResponse<CowSwapQuoteResponse> =
+      await cowService.post<CowSwapQuoteResponse>(`${deps.apiUrl}/v1/quote/`, {
+        sellToken: sellAssetErc20Address,
+        buyToken: buyAssetErc20Address,
+        receiver: receiveAddress,
+        validTo: getNowPlusThirtyMinutesTimestamp(),
+        appData: DEFAULT_APP_DATA,
+        partiallyFillable: false,
+        from: receiveAddress,
+        kind: ORDER_KIND_SELL,
+        sellAmountBeforeFee: normalizedSellAmount
+      })
+
+    const {
+      data: { quote }
+    } = quoteResponse
+
+    const rate = bn(quote.buyAmount)
+      .div(quote.sellAmount)
+      .times(bn(10).exponentiatedBy(sellAsset.precision - buyAsset.precision))
+      .toString()
+
+    const [feeDataOptions, wethUsdRate, usdRateSellAsset] = await Promise.all([
+      adapter.getFeeData({
+        to: COW_SWAP_VAULT_RELAYER_ADDRESS,
+        value: '0',
+        chainSpecific: { from: receiveAddress, contractAddress: sellAssetErc20Address }
+      }),
+      getUsdRate(deps, feeAsset),
+      getUsdRate(deps, sellAsset)
+    ])
+    const feeData = feeDataOptions['fast']
+
+    const fee = bnOrZero(quote.feeAmount)
```

Comment:
similarly here, can we break up into a few lines? this logic is pretty terse.

Reply:
@0xdef1cafe agreed should have done that earlier, split it into more comprehensive code and added comments !